cpp/arcticdb/version/version_core.cpp

```diff
@@ -819,6 +819,43 @@ void read_incompletes_to_pipeline(
     pipeline_context->total_rows_ = pipeline_context->calc_rows();
 }
 
+void check_incompletes_index_ranges_dont_overlap(const std::shared_ptr<PipelineContext>& pipeline_context) {
+    // Does nothing if the symbol is not timestamp-indexed
+    // Checks both that the index ranges of incomplete segments do not overlap with one another, and that the earliest
+    // timestamp in an incomplete segment is greater than the latest timestamp existing in the symbol in the case of a
+    // parallel append
+    if (pipeline_context->descriptor().index().type() == IndexDescriptor::TIMESTAMP) {
+        std::optional<timestamp> last_existing_index_value;
+        // Beginning of incomplete segments == beginning of all segments implies all segments are incompletes, so we are
+        // writing, not appending
+        if (pipeline_context->incompletes_begin() != pipeline_context->begin()) {
+            auto last_indexed_slice_and_key = std::prev(pipeline_context->incompletes_begin())->slice_and_key();
+            // -1 as end_time is stored as 1 greater than the last index value in the segment
+            last_existing_index_value = last_indexed_slice_and_key.key().end_time() - 1;
+        }
+
+        // Use ordered set so we only need to compare adjacent elements
+        std::set<TimestampRange> unique_timestamp_ranges;
+        for (auto it = pipeline_context->incompletes_begin(); it!= pipeline_context->end(); it++) {
+            sorting::check<ErrorCode::E_UNSORTED_DATA>(
+                    !last_existing_index_value.has_value() || it->slice_and_key().key().start_time() > *last_existing_index_value,
+                    "Cannot append staged segments to existing data as incomplete segment contains index value <= existing data: {} <= {}",
+                    it->slice_and_key().key().start_time(),
+                    *last_existing_index_value);
+            unique_timestamp_ranges.insert({it->slice_and_key().key().start_time(), it->slice_and_key().key().end_time()});
+        }
+        for (auto it = unique_timestamp_ranges.begin(); it != unique_timestamp_ranges.end(); it++) {
+            auto next_it = std::next(it);
+            if (next_it != unique_timestamp_ranges.end()) {
+                sorting::check<ErrorCode::E_UNSORTED_DATA>(
+                        next_it->first >= it->second,
+                        "Cannot append staged segments to existing data as incomplete segment index values overlap one another: ({}, {}) ∩ ({}, {}) ≠ ∅",
```

Comment:
Maybe change the notation to be a bit more friendly?

Reply:
I suppose enforcing all users to have basic set-theory to understand our error messages might be exclusionary